reconciletags/reconciletags.py

```diff
@@ -23,87 +23,99 @@
+                        as the image can't be found""".format(full_digest))
         return existing_tags
 
-    def get_latest_digest(self, repo):
-        output = json.loads(self.call('gcloud container images list-tags '
-                            '--no-show-occurrences {0}'.format(repo), False))
-        for image in output:
-            if 'latest' in image['tags']:
-                return image['digest']
+    def get_latest_digest(self, manifests):
+        for digest in manifests:
+            if 'latest' in manifests[digest]['tag']:
+                return digest
 
     def reconcile_tags(self, data, dry_run):
-        # Hardcode dry_run to False for this call because we always want
-        # want to see config regardless of whether we actually run the
-        # reconciler.
-        self.call('gcloud config list', False)
         for project in data['projects']:
             default_registry = project['base_registry']
-            # additional registries are optional, just default to an empty list
-            # if it's absent from the config
+
             registries = project.get('additional_registries', [])
             registries.append(default_registry)
             for registry in registries:
                 full_repo = os.path.join(registry, project['repository'])
-                default_repo = os.path.join(default_registry,
-                                            project['repository'])
-                existing_tags = self.get_existing_tags(full_repo)
-                latest = self.get_latest_digest(full_repo)
-                logging.debug(existing_tags)
 
                 for image in project['images']:
-                    full_digest = default_repo + '@sha256:' + image['digest']
+                    full_digest = full_repo + '@sha256:' + image['digest']
                     full_tag = full_repo + ':' + image['tag']
 
-                    # Don't retag latest if it's already latest
-                    if latest:
-                        if latest.startswith('sha256:'):
-                            latest = latest[len('sha256:'):]
-                        if (image['tag'] == 'latest'
-                           and latest.startswith(image['digest'])):
-                            logging.debug('Skipping tagging %s as latest as '
-                                          'it is already latest.',
-                                          image['digest'])
-                            continue
-
-                    self.add_tags(full_digest, full_tag, dry_run)
-
-                logging.debug(self.get_existing_tags(full_repo))
+                    name = docker_name.Digest(full_digest)
+                    creds = docker_creds.DefaultKeychain.Resolve(name)
+                    transport = transport_pool.Http(httplib2.Http)
+
+                    with docker_image.FromRegistry(
+                        name, creds, transport) as img:
+                            if img.exists():
```

Comment:
To make this a little bit less nested, consider factoring out the logic for a single image into a function.

You can also do something like:

```python
if not img.exists():
  logging.warn()
  return
do other stuff
```